File: packaging/macos/ConduitGUI/Sources/ConduitGUI/Consent.swift
import AppKit
import SwiftUI

/// One floating window shows queued consent requests one at a time. Deny is the
/// default; Allow arms only after a short delay so a stray click or key-repeat
/// can't approve anything. Mirrors gui-gtk/src/consent.rs.
final class ConsentState: ObservableObject {
    @Published var current: JSON = .null
    @Published var allowArmed = false
    var onAnswer: ((_ allow: Bool, _ perms: [String], _ remember: Bool, _ scope: String, _ path: String) -> Void)?
}

final class ConsentController: NSObject, NSWindowDelegate {
    private let model: AppModel
    private let state = ConsentState()
    private var window: NSWindow?
    private var queue: [JSON] = []
    private var currentId: UInt64?
    private var armTimer: Timer?

    init(model: AppModel) {
        self.model = model
        super.init()
        state.onAnswer = { [weak self] allow, perms, remember, scope, path in
            self?.answer(allow: allow, perms: perms, remember: remember, scope: scope, path: path)
        }
    }

    private func id(of req: JSON) -> UInt64 { UInt64(req["id"].int64 ?? 0) }

    func add(_ req: JSON) {
        let rid = id(of: req)
        if currentId == rid || queue.contains(where: { id(of: $0) == rid }) { return } // dedup
        queue.append(req)
        if currentId == nil { showNext() }
    }

    func remove(_ rid: UInt64) {
        queue.removeAll { id(of: $0) == rid }
        if currentId == rid { currentId = nil; showNext() }
    }

    private func answer(allow: Bool, perms: [String], remember: Bool, scope: String, path: String) {
        guard let rid = currentId else { return }
        model.cmd("consent", [
            "id": .number(Double(rid)),
            "allow": .bool(allow),
            "perms": .array(perms.map { .string($0) }),
            "remember": .bool(remember),
            "scope": .string(scope),
            "path": .string(path),
        ])
        currentId = nil
        showNext()
    }

    private func showNext() {
        armTimer?.invalidate()
        guard !queue.isEmpty else { window?.orderOut(nil); return }
        let req = queue.removeFirst()
        currentId = id(of: req)
        state.current = req
        state.allowArmed = false
        ensureWindow()
        window?.makeKeyAndOrderFront(nil)
        NSApp.activate(ignoringOtherApps: true)
        armTimer = Timer.scheduledTimer(withTimeInterval: 0.7, repeats: false) { [weak self] _ in
            self?.state.allowArmed = true
        }
    }

    private func ensureWindow() {
        if window != nil { return }
        let host = NSHostingController(rootView: ConsentView(state: state))
        let w = NSWindow(contentViewController: host)
        w.title = Loc.t("app")
        w.styleMask = [.titled, .closable]
        w.level = .floating
        w.isReleasedWhenClosed = false
        w.delegate = self
        w.center()
        window = w
    }

    // Clicking the window's close button denies whatever is showing; the window
    // itself is kept for the next prompt (showNext hides it when the queue empties).
    func windowShouldClose(_ sender: NSWindow) -> Bool {
        answer(allow: false, perms: [], remember: false, scope: "", path: "")
        return false
    }
}

private struct ConsentView: View {
    @ObservedObject var state: ConsentState
    @State private var perms: [String: Bool] = [:]
    @State private var remember = false
    @State private var scope = "always"

    private let scopes: [(String, String)] = [
        ("session", "scope_session"), ("1h", "scope_1h"), ("1d", "scope_1d"), ("always", "scope_always"),
    ]

    var body: some View {
        let req = state.current
        let origin = req["origin"].string ?? ""
        let kind = req["kind"].string ?? ""
        let detail = req["detail"].string ?? ""
        let pairPerms = req["perms"].array.compactMap { $0.string }

        return VStack(alignment: .leading, spacing: 12) {
            Text(origin).font(.title3.weight(.semibold)).fixedSize(horizontal: false, vertical: true)
            Text(message(kind: kind, origin: origin, detail: detail))
                .fixedSize(horizontal: false, vertical: true)

            if kind == "pair" {
                VStack(alignment: .leading, spacing: 4) {
                    ForEach(pairPerms, id: \.self) { p in
                        Toggle(Loc.t("perm_" + p), isOn: Binding(
                            get: { perms[p] ?? true },
                            set: { perms[p] = $0 }
                        ))
                    }
                }
                Picker(Loc.t("grant_for"), selection: $scope) {
                    ForEach(scopes, id: \.0) { code, key in Text(Loc.t(key)).tag(code) }
                }
            }
            if kind == "pair" || kind == "launch" {
                Toggle(Loc.t("remember"), isOn: $remember)
            }

            HStack {
                Spacer()
                Button(Loc.t("deny")) { state.onAnswer?(false, [], false, "", "") }
                    .keyboardShortcut(.cancelAction)
                if kind == "folder" {
                    // A folder request is answered by choosing a folder.
                    Button(Loc.t("choose_folder")) { chooseFolder() }
                        .keyboardShortcut(.defaultAction)
                        .disabled(!state.allowArmed)
                } else {
                    Button(Loc.t("allow")) {
                        let picked = kind == "pair" ? pairPerms.filter { perms[$0] ?? true } : []
                        state.onAnswer?(true, picked, remember, kind == "pair" ? scope : "", "")
                    }
                    .keyboardShortcut(.defaultAction)
                    .disabled(!state.allowArmed)
                }
            }
        }
        .padding(20)
        .frame(width: 420)
        .environment(\.layoutDirection, Loc.rtl ? .rightToLeft : .leftToRight)
        .onChange(of: idValue) { _ in perms = [:]; remember = false; scope = "always" }
    }

    private func chooseFolder() {
        let panel = NSOpenPanel()
        panel.canChooseFiles = false
        panel.canChooseDirectories = true
        panel.allowsMultipleSelection = false
        panel.prompt = Loc.t("choose_folder")
        if panel.runModal() == .OK, let url = panel.url {
            state.onAnswer?(true, [], false, "", url.path)
        }
    }

    private var idValue: Double { Double(state.current["id"].int64 ?? 0) }

    private func message(kind: String, origin: String, detail: String) -> String {
        switch kind {
        case "pair": return Loc.f("consent_pair", ["origin": origin])
        case "launch": return Loc.f("consent_launch", ["origin": origin]) + "\n" + detail
        case "kill": return Loc.f("consent_kill", ["origin": origin]) + "\n" + detail
        case "power": return Loc.t("consent_power") + " " + detail
        case "clipboard": return Loc.f("consent_clipboard", ["origin": origin])
        case "elevate": return Loc.f("consent_elevate", ["origin": origin])
        case "hostwrite": return Loc.f("consent_hostwrite", ["origin": origin, "verb": detail])
        default: return kind + " " + detail
        }
    }
}

File: gui-gtk/src/consent.rs
//! The consent prompt. One window shows queued requests one at a time. Deny is
//! always available and is the default; Allow arms after a short delay so a
//! stray click or key-repeat can't approve anything. Pairing requests show a
//! checkbox per permission and a "remember" option.

use crate::bus::Bus;
use crate::loc::{t, tf};
use gtk::prelude::*;
use gtk::{glib, Align, Application, Box as GBox, Button, CheckButton, Label, Orientation, Window};
use serde_json::{json, Value};
use std::cell::RefCell;
use std::rc::Rc;
use std::time::Duration;

struct Inner {
    window: Window,
    body: GBox,
    queue: Vec<Value>,
    current: Option<u64>,
    bus: Bus,
}

#[derive(Clone)]
pub struct ConsentWindow {
    inner: Rc<RefCell<Inner>>,
}

impl ConsentWindow {
    pub fn new(app: &Application, bus: &Bus) -> Self {
        let window = Window::builder()
            .application(app)
            .title(&t("app"))
            .default_width(440)
            .modal(true)
            .resizable(false)
            .build();
        if crate::loc::is_rtl() {
            window.set_direction(gtk::TextDirection::Rtl);
        }
        let body = GBox::new(Orientation::Vertical, 12);
        body.set_margin_top(20);
        body.set_margin_bottom(20);
        body.set_margin_start(24);
        body.set_margin_end(24);
        window.set_child(Some(&body));

        let this = ConsentWindow { inner: Rc::new(RefCell::new(Inner { window, body, queue: Vec::new(), current: None, bus: bus.clone() })) };
        // Closing the window denies whatever is showing.
        let w = this.clone();
        this.inner.borrow().window.connect_close_request(move |_| {
            w.answer(false, vec![], false, "", "");
            gtk::Inhibit(false)
        });
        this
    }

    pub fn add(&self, req: &Value) {
        let id = req["id"].as_u64().unwrap_or(0);
        {
            let mut i = self.inner.borrow_mut();
            if i.current == Some(id) || i.queue.iter().any(|r| r["id"].as_u64() == Some(id)) {
                return; // dedup: a restarted GUI may resend queued prompts
            }
            i.queue.push(req.clone());
        }
        if self.inner.borrow().current.is_none() {
            self.show_next();
        }
    }

    pub fn remove(&self, id: u64) {
        let advance = {
            let mut i = self.inner.borrow_mut();
            i.queue.retain(|r| r["id"].as_u64() != Some(id));
            i.current == Some(id)
        };
        if advance {
            self.inner.borrow_mut().current = None;
            self.show_next();
        }
    }

    fn answer(&self, allow: bool, perms: Vec<String>, remember: bool, scope: &str, path: &str) {
        let id = {
            let mut i = self.inner.borrow_mut();
            let Some(id) = i.current.take() else { return };
            i.bus.send(json!({"cmd": "consent", "id": id, "allow": allow, "perms": perms,
                              "remember": remember, "scope": scope, "path": path}));
            id
        };
        let _ = id;
        self.show_next();
    }

    fn show_next(&self) {
        let req = {
            let mut i = self.inner.borrow_mut();
            if i.current.is_some() {
                return;
            }
            match i.queue.first().cloned() {
                Some(r) => {
                    i.current = r["id"].as_u64();
                    i.queue.remove(0);
                    r
                }
                None => {
                    i.window.set_visible(false);
                    return;
                }
            }
        };
        self.render(&req);
    }

    fn render(&self, req: &Value) {
        let (body, window) = {
            let i = self.inner.borrow();
            (i.body.clone(), i.window.clone())
        };
        while let Some(c) = body.first_child() {
            body.remove(&c);
        }

        let origin = req["origin"].as_str().unwrap_or("").to_string();
        let kind = req["kind"].as_str().unwrap_or("").to_string();
        let detail = req["detail"].as_str().unwrap_or("").to_string();

        let title = Label::new(Some(&origin));
        title.set_xalign(0.0);
        title.add_css_class("title-3");
        title.set_wrap(true);
        body.append(&title);

        let o = [("origin", origin.as_str())];
        let ask = match kind.as_str() {
            "pair" => tf("consent_pair", &o),
            "launch" => format!("{}\n{detail}", tf("consent_launch", &o)),
            "kill" => format!("{}\n{detail}", tf("consent_kill", &o)),
            "power" => format!("{} {detail}", t("consent_power")),
            "clipboard" => tf("consent_clipboard", &o),
            "elevate" => tf("consent_elevate", &o),
            "hostwrite" => tf("consent_hostwrite", &[("origin", origin.as_str()), ("verb", detail.as_str())]),
            "folder" => tf("consent_folder", &o),
            "shell" => format!("{}\n{detail}", tf("consent_shell", &o)),
            _ => format!("{kind} {detail}"),
        };
        let msg = Label::new(Some(&ask));
        msg.set_xalign(0.0);
        msg.set_wrap(true);
        body.append(&msg);

        // Pairing: a checkbox per requested permission.
        let perm_checks: Rc<RefCell<Vec<(String, CheckButton)>>> = Rc::new(RefCell::new(Vec::new()));
        if kind == "pair" {
            if let Some(perms) = req["perms"].as_array() {
                let list = GBox::new(Orientation::Vertical, 8);
                list.set_margin_top(10);
                list.set_margin_bottom(6);
                for p in perms {
                    let name = p.as_str().unwrap_or("").to_string();
                    let cb = CheckButton::with_label(&t(&format!("perm_{name}")));
                    cb.set_active(true);
                    cb.set_margin_top(2);
                    cb.set_margin_bottom(2);
                    list.append(&cb);
                    perm_checks.borrow_mut().push((name, cb));
                }
                body.append(&list);
                // One click to (re)check every requested permission.
                let grant_all = Button::with_label(&t("grant_all"));
                grant_all.add_css_class("flat");
                grant_all.set_halign(Align::Start);
                let pc = perm_checks.clone();
                grant_all.connect_clicked(move |_| {
                    for (_, cb) in pc.borrow().iter() {
                        cb.set_active(true);
                    }
                });
                body.append(&grant_all);
            }
        }

        // Pairing: choose how long the grant lasts.
        let scope_codes = ["session", "1h", "1d", "always"];
        let scope_dd = gtk::DropDown::from_strings(&[
            &t("scope_session"), &t("scope_1h"), &t("scope_1d"), &t("scope_always"),
        ]);
        scope_dd.set_selected(3); // default: always
        if kind == "pair" {
            let row = GBox::new(Orientation::Horizontal, 8);
            row.set_margin_top(6);
            let lbl = Label::new(Some(&t("grant_for")));
            lbl.set_xalign(0.0);
            row.append(&lbl);
            row.append(&scope_dd);
            body.append(&row);
        }

        // "Remember" for the kinds that support it.
        let remember = CheckButton::with_label(&t("remember"));
        if matches!(kind.as_str(), "pair" | "launch") {
            body.append(&remember);
        }

        let buttons = GBox::new(Orientation::Horizontal, 8);
        buttons.set_halign(Align::End);
        buttons.set_margin_top(8);
        let deny = Button::with_label(&t("deny"));
        buttons.append(&deny);

        let this = self.clone();
        deny.connect_clicked(move |_| this.answer(false, vec![], false, "", ""));

        if kind == "folder" {
            // A folder request is answered by choosing a folder, not by "Allow".
            let choose = Button::with_label(&t("choose_folder"));
            choose.add_css_class("suggested-action");
            buttons.append(&choose);
            let this = self.clone();
            let win = window.clone();
            choose.connect_clicked(move |_| {
                let chooser = gtk::FileChooserNative::new(
                    Some(&t("choose_folder")),
                    Some(&win),
                    gtk::FileChooserAction::SelectFolder,
                    Some(&t("choose_folder")),
                    Some(&t("cancel")),
                );
                let ch = chooser.clone();
                let this = this.clone();
                chooser.connect_response(move |_, resp| {
                    if resp == gtk::ResponseType::Accept {
                        if let Some(p) = ch.file().and_then(|f| f.path()) {
                            this.answer(true, vec![], false, "", p.to_string_lossy().as_ref());
                        }
                    }
                    ch.destroy();
                });
                chooser.show();
            });
        } else {
            let allow = Button::with_label(&t("allow"));
            allow.add_css_class("suggested-action");
            allow.set_sensitive(false); // armed after a short delay
            buttons.append(&allow);
            let allow2 = allow.clone();
            glib::timeout_add_local_once(Duration::from_millis(700), move || allow2.set_sensitive(true));

            let this = self.clone();
            let pc = perm_checks.clone();
            let rem = remember.clone();
            let kind2 = kind.clone();
            allow.connect_clicked(move |_| {
                let perms: Vec<String> = if kind2 == "pair" {
                    pc.borrow().iter().filter(|(_, c)| c.is_active()).map(|(n, _)| n.clone()).collect()
                } else {
                    req_perms(&kind2)
                };
                let scope = if kind2 == "pair" {
                    scope_codes.get(scope_dd.selected() as usize).copied().unwrap_or("always")
                } else {
                    ""
                };
                this.answer(true, perms, rem.is_active(), scope, "");
            });
        }
        body.append(&buttons);

        window.present();
    }
}

/// Non-pair kinds don't send a permission list; the core knows the scope.
fn req_perms(_kind: &str) -> Vec<String> {
    Vec::new()
}
